### PR TITLE
GEOMESA-2698 SimpleFeatureType config to prioritize time-based indices

### PR DESCRIPTION
**Pull request by elahrvivaz**
Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: docs/user/datastores/index_config.rst [hunk 1]
```diff
@@ -649,6 +649,19 @@ For example:
 
 See :ref:`cli_analytic` and :ref:`stats_api` for information on reading cached stats.
 
+Configuring Temporal Priority
+-----------------------------
+
+For some large time-based datasets, an index that leverages a temporal predicate will almost always be faster
```
Comment (jnh5y) [inline docs/user/datastores/index_config.rst:655]: If we believe this, should we make the default 'true'?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/FilterSplitter.scala [hunk 2]
```diff
@@ -285,7 +286,9 @@ object FilterSplitter {
     }
 
     if (canMergePrimary && toMerge.secondary == mergeTo.secondary) {
-      FilterStrategy(mergeTo.index, orOption(toMerge.primary.toSeq ++ mergeTo.primary), mergeTo.secondary, mergeTo.cost)
+      val primary = orOption(toMerge.primary.toSeq ++ mergeTo.primary)
+      val temporal = mergeTo.temporal && toMerge.temporal // should be the same...
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/FilterSplitter.scala:290]: One can only hope.:)

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/StrategyDecider.scala [hunk 3]
```diff
@@ -131,23 +131,39 @@ object StrategyDecider extends MethodProfiling with LazyLogging {
...
+
+    private def cost(option: FilterPlan): FilterPlanCost = {
+      profile((fpc: FilterPlanCost, time: Long) => fpc.copy(time = time)) {
+        FilterPlanCost(option, option.strategies.map(_.cost).sum, 0L)
+      }
+    }
+
+    private case class FilterPlanCost(plan: FilterPlan, cost: Long, time: Long) extends Comparable[FilterPlanCost] {
+      override def compareTo(o: FilterPlanCost): Int = java.lang.Long.compare(cost, o.cost)
+      override def toString: String = s"$plan (Cost $cost in ${time}ms)"
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/StrategyDecider.scala:166]: Interesting... are we seeing cost calculations taking a noticeable amount of time?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/IdFilterStrategy.scala [hunk 4]
```diff
@@ -17,20 +17,22 @@ import org.opengis.filter.{And, Filter, Id, Or}
...
+      Some(FilterStrategy(this, None, None, temporal = false, Long.MaxValue))
     } else if (filter == Filter.EXCLUDE) {
       None
     } else {
       val (ids, notIds) = IdExtractingVisitor(filter)
       if (ids.isDefined) {
         // top-priority index - always 1 if there are actually ID filters
-        Some(FilterStrategy(this, ids, notIds, IdFilterStrategy.StaticCost))
+        // note: although there's no temporal predicate, there's an implied exact date for the given feature
+        Some(FilterStrategy(this, ids, notIds, temporal = true, IdFilterStrategy.StaticCost))
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/IdFilterStrategy.scala:33]: Nice corner case.  

This does seem to suggest that we may want to think through all the cases our query planning handles at some point and think through it holistically.  

It seems good and bad to associate an id query (which has the ids specified) as 'temporal'.  Locally, right now, it is the Right Thing though.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/SpatialFilterStrategy.scala [hunk 5]
```diff
@@ -19,23 +19,24 @@ trait SpatialFilterStrategy[T, U] extends GeoMesaFeatureIndex[T, U] {
 
   import SpatialFilterStrategy.{StaticCost, spatialCheck}
 
-  def geom: String
+  // attributes are assumed to be a single geometry field
+  lazy private val Seq(geom) = attributes
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/SpatialFilterStrategy.scala:23]: Is there any way that this is not true?

Comment (jnh5y) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/SpatialFilterStrategy.scala:23]: I'm just wondering if there's any chance for an exception, etc.  (It seems believable that everything is ok here; just asking.)

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStoreTest.scala [hunk 6]
```diff
@@ -235,6 +237,24 @@ class GeoMesaDataStoreTest extends Specification {
...
+      val spec = "name:String:index=true,age:Int,dtg:Date,*geom:Point:srid=4326"
+
+      val ds = new TestGeoMesaDataStore(true)
+      ds.createSchema(SimpleFeatureTypes.createType("default", spec))
+      ds.createSchema(SimpleFeatureTypes.createType("temporal", s"$spec;geomesa.temporal.priority=true"))
+
+      val filters = Seq(
+        ("name like 'a%' and dtg DURING 2020-01-01T00:00:00.00Z/2020-01-02T00:00:00.00Z", AttributeIndex, Z3Index),
+        ("name > 'a' AND name < 'b' and dtg DURING 2020-01-01T00:00:00.00Z/2020-01-02T00:00:00.00Z", AttributeIndex, Z3Index),
+        ("name = 'a' and dtg DURING 2020-01-01T00:00:00.00Z/2020-01-02T00:00:00.00Z", AttributeIndex, AttributeIndex)
```
Comment (jnh5y) [inline geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStoreTest.scala:250]: Would it be 'easy' to add tests for the id cases above?

File: docs/user/datastores/index_config.rst (hunk 1 above)
Comment (elahrvivaz) [reply docs/user/datastores/index_config.rst:655]: well, there's the `For some large time-based datasets` part...

Comment (elahrvivaz) [reply docs/user/datastores/index_config.rst:655]: in practice, i think the only thing this affects is prefix attribute scans against high-cardinality attributes get de-prioritized below z3.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/StrategyDecider.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/StrategyDecider.scala:166]: not with the current heuristic impl - using stats could take a bit longer. we've always had this logging in here, i just cleaned it up a little.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/IdFilterStrategy.scala (hunk 4 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/IdFilterStrategy.scala:33]: yeah, the current approach of using costs and delegating the cost to the index means that we don't have a single place to rank everything... but the tradeoff is that it's extensible.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/SpatialFilterStrategy.scala (hunk 5 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/strategies/SpatialFilterStrategy.scala:23]: only if there's a new class that extends this trait and doesn't follow the current conventions... i thought about adding some validation here but seemed like overkill.

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStoreTest.scala (hunk 6 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStoreTest.scala:250]: easy enough - but id filters are always #1 priority so it wouldn't show any change. i guess that would be a meaningful result in itself. i'll add one

Comment (elahrvivaz) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStoreTest.scala]: added

**Comment by jnh5y** [comment]
Merged as c11aa3c279e1c7106dbaf8864c5b1dd17a658951

**Pull request closed without merging**